Support wasm

danielgranhao opened the pull request:
Resolves #90.

This PR:

- Changes the http client from `ureq` to async `reqwest`
- Switches to a different secp256k1-zkp fork that circumvents wasm compatibility issues
- Changes the websocket crate from `tungstenite` to `tokio-tungstenite-wasm` - as the name indicates, this crate uses `tokio-tungstenite` for native builds
- Adds esplora as a supported chain data backend. Electrum is still supported. Both can be enabled using features, allowing clients to switch backends at runtime (which may be useful for redundancy). Only esplora is set as a default so that wasm is supported by default.
- Allows reusing esplora and electrum clients instead of having to create new ones on every method.

~~There are still some TODOs, but I would appreciate some initial feedback.~~

TODO:

- [x] Testing (adding new tests for esplora + run ignored tests)
    - I've made the testnet and mainnet integration tests run against regtest and added [regtest](https://github.com/BoltzExchange/regtest) as a submodule. These tests now run in CI both on native and wasm builds. 
- [x] Try to upstream changes to secp256k1-zkp
    - I've rebased[ sanket's PR](https://github.com/BlockstreamResearch/rust-secp256k1-zkp/pull/48) (on which we depended already) on top of the secp256k1-zkp main branch (to pull in a [WASM fix](https://github.com/BlockstreamResearch/rust-secp256k1-zkp/pull/86)) and added an additional [workaround/fix](https://github.com/danielgranhao/rust-secp256k1-zkp/commit/eac2e479255a6e32b5588bc25ee53c642fdd8395) on top. I won't be trying to upstream the changes because his PR has been dropped in favor of adding musig2 support on the base secp256k1 crate (WIP [here](https://github.com/rust-bitcoin/rust-secp256k1/pull/716)). Hopefully soon a fork won't be needed any longer.

danielgranhao committed:
Fix secp256k1-zkp WASM compatibility

danielgranhao committed:
Enable js feature in getrandom

danielgranhao committed:
Use async reqwest

danielgranhao committed:
Use tokio-tungstenite-wasm for websocket stream

danielgranhao committed:
Add esplora support

danielgranhao committed:
Adjust GH actions

danielgranhao committed:
Update README.md

danielgranhao committed:
Fix esplora timeout const name

danielgranhao committed:
Esplora fixes

danielgranhao committed:
Use test macros where missing

danielgranhao committed:
Add missing esplora conditionals

danielgranhao committed:
Fix get swap

roeierez commented on src/swaps/liquid.rs:
```diff
@@ -404,37 +401,19 @@ impl LBtcSwapScript {
-        let electrum_client = network_config.clone().build_client()?;
+    pub async fn fetch_utxo<LC: LiquidClient, N: LiquidNetworkConfig<LC>>(
+        &self,
+        network_config: &N,
```
Passing a config forces this function to call `build_client` (and create a new client) which prevents from the caller to reuse a client implementation. In electrum we have seen it has a big performance penalty.
I know this was the implementation before this PR but perhaps if we are going for such a big change we can improve it?
I commented on this method as an example but this pattern repeats itself.

danielgranhao replied:
Thanks @roeierez. I agree. I've made the methods take a client instead of the config in [a6d9d1c](https://github.com/SatoshiPortal/boltz-rust/pull/91/commits/a6d9d1c54128ef6edbbf4ba2b4edc2f41f5c629f)

danielgranhao committed:
Fix fmt lint

danielgranhao committed:
Run tests on regtest

danielgranhao committed:
Update secp256k1-zkp patch

danielgranhao committed:
Allow chain client reuse

danielgranhao committed:
Update readme

danielgranhao committed:
Rename const

michael1011 commented on regtest/proxy/ssl_proxy.py:
```diff
@@ -0,0 +1,106 @@
+from flask import Flask, request, json, Response
```
What do we need that proxy for?

danielgranhao replied:
The browser is very picky when it comes to tls certificates (doesn't like self-signed ones) as well as CORS. The proxy allows circumventing those issues when connecting to the regtest LND, bitcoind and elementsd rpc interfaces when running WASM integration tests.

michael1011 commented on src/network/electrum.rs:
```diff
@@ -128,32 +127,187 @@ impl ElectrumConfig {
+                        (
+                            bitcoin::OutPoint::new(tx.compute_txid(), vout as u32),
+                            output.clone(),
+                        )
```
Can we use `into_iter` to avoid the clone?

danielgranhao replied:
Optimized the method in [2fb92c9](https://github.com/SatoshiPortal/boltz-rust/pull/91/commits/2fb92c9e2238d891cf7706880a99f2ce996addf1)

michael1011 commented on src/network/electrum.rs:
```diff
@@ -128,32 +127,187 @@ impl ElectrumConfig {
+    async fn get_address_balance(&self, address: &Address) -> Result<(u64, i64), Error> {
+        let spk = address.script_pubkey();
+        let script_balance = self.inner.script_get_balance(spk.as_script())?;
+        Ok((script_balance.confirmed, script_balance.unconfirmed))
```
Just out of curiosity: why is `script_balance.unconfirmed` an `i64`?

danielgranhao replied:
Some servers (like electrs) return a negative value to imply that funds are being spent.

michael1011 commented on src/network/electrum.rs:
```diff
@@ -128,32 +127,187 @@ impl ElectrumConfig {
+    async fn broadcast_tx(&self, signed_tx: &Transaction) -> Result<Txid, Error> {
+        Ok(self.inner.transaction_broadcast(signed_tx)?)
+    }
+    fn network(&self) -> Chain {
```
```suggestion

    fn network(&self) -> Chain {
```

michael1011 reviewed (commented):
Untested concept ACK

danielgranhao committed:
Switch to vendored native tls on tokio-tungetenite

danielgranhao committed:
Expose WebSocketStream

roeierez commented on src/network/electrum.rs:
```diff
@@ -128,32 +127,187 @@ impl ElectrumConfig {
+        })
+    }
+
+    fn fetch_utxos_core(
```
NIT: fetch implies some IO like network request. Perhaps we can use `extract_script_utxos`?

danielgranhao replied:
Changed the name to `extract_address_utxos` in [58f93e5](https://github.com/SatoshiPortal/boltz-rust/pull/91/commits/58f93e5f6d974136f709e93b7fbebd4277f67c76)

roeierez commented on src/network/electrum.rs:
```diff
@@ -128,32 +127,187 @@ impl ElectrumConfig {
-    // Get a copy of the network (Chain) field.
-    pub fn network(&self) -> Chain {
+
+    pub fn build_bitcoin_client(&self) -> Result<ElectrumBitcoinClient, Error> {
+        ElectrumBitcoinClient::new(self.url.clone(), self.timeout, self.network)
+    }
+
+    pub fn build_liquid_client(&self) -> Result<ElectrumLiquidClient, Error> {
+        ElectrumLiquidClient::new(self.url.clone(), self.timeout, self.network)
+    }
```
Not sure we need these functions on the config. It looks like it is possible for users to build the wrong client without any error.
Do you see a drawback for removing these methods? The specific client creation looks pretty simple even without them.

danielgranhao replied:
I think the "factory-like" aspect of them provides convenience, but I agree they aren't essential. Maybe the root issue is having a single config struct for both networks? I could split the config struct into Bitcoin and Liquid counterparts

roeierez replied:
@danielgranhao  It seem to me that the ElectrumConfig is only used to create the client. Perhaps we can get rid of the config now that we have clients?

danielgranhao replied:
Done in [bd904fc](https://github.com/SatoshiPortal/boltz-rust/pull/91/commits/bd904fcef332360899fcfbf1b2e237977f80f864) - (edit: I was referring to the previous comment)

danielgranhao replied:
You're right, thanks. Removed it in [73a41a3](https://github.com/SatoshiPortal/boltz-rust/pull/91/commits/73a41a3d3f2a08b6442f57f77598cc00aa639429)

roeierez reviewed (commented):
Looks good! Added two NITs.

danielgranhao committed:
Optimize electrum's fetch_utxos_core

danielgranhao committed:
Add missing new line

Co-authored-by: michael1011 <me@michael1011.at>

danielgranhao committed:
Rename fetch_utxo_core to extract_address_utxos

danielgranhao committed:
Split configs and chain into Bitcoin and Liquid counterparts

danielgranhao committed:
Expose entire tokio-tungstenite-wasm

michael1011 commented on tests/regtest/mod.rs:
```diff
@@ -0,0 +1,6 @@
+mod chain_swaps;
+mod liquid;
+
+const WAIT_TIME_MS: i32 = 5_000;
```
Just fyi: I guess we can remove or reduce that value after https://github.com/BoltzExchange/boltz-backend/pull/848 is merged

danielgranhao committed:
Drop client configs

danielgranhao commented:
There is an issue with the websocket client. `tokio-tungstenite-wasm` doesn't expose websocket pings because they are not available in the javascript API. IIUC some browsers handle it automatically. This wrapper crate tries to keep the API consistent between native and WASM, so it also hides the ping-pong API for native builds. 

I see two main solution paths:

-  Fork the `tokio-tungstenite-wasm` in order to expose the ping API on native (eventually try to upstream it). Not sure how viable this is. Also, I'm not sure how reliable the auto-keep-alive implementation of the browsers is. 
- Implement a keep-alive ping-pong on the app level. This could be as simple as the text "ping" and "pong" but it requires adapting the boltz backend accordingly cc @michael1011. This is the safest option IMO and the overhead should be insignificant.

michael1011 commented:
> There is an issue with the websocket client. `tokio-tungstenite-wasm` doesn't expose websocket pings because they are not available in the javascript API. IIUC some browsers handle it automatically. This wrapper crate tries to keep the API consistent between native and WASM, so it also hides the ping-pong API for native builds.
> 
> I see two main solution paths:
> 
> * Fork the `tokio-tungstenite-wasm` in order to expose the ping API on native (eventually try to upstream it). Not sure how viable this is. Also, I'm not sure how reliable the auto-keep-alive implementation of the browsers is.
> * Implement a keep-alive ping-pong on the app level. This could be as simple as the text "ping" and "pong" but it requires adapting the boltz backend accordingly cc @michael1011. This is the safest option IMO and the overhead should be insignificant.

I'd prefer a ping/pong JSON message in the application layer over having to fork

danielgranhao commented:
I opened a [PR](https://github.com/BoltzExchange/boltz-backend/pull/850) on boltz-backend to add these ping/pong messages.

danielgranhao committed:
Refactor ws message structs and add ping/pong messages

danielgranhao commented:
In [f48f3d5](https://github.com/SatoshiPortal/boltz-rust/pull/91/commits/f48f3d514f7485191b67ceca563caf7de9d58d45) I've adjusted the ws message structs to align with https://github.com/BoltzExchange/boltz-backend, including the newly added ping/pong messages. Added also regtest test for the pinging. 

This is ready for another review round. cc @michael1011 @i5hi @roeierez

dangeross commented on Cargo.toml:
```diff
@@ -4,13 +4,18 @@ description = "a boltz exchange client for swaps between BTC/LBTC & LN"
 authors = ["i5hi <ishi@satoshiportal.com>", "Rajarshi Maitra <raj@bitshala.org>"]
 version = "0.2.0"
```
Might be worth bumping the version now the WS message structs have changed

danielgranhao replied:
Done [662ab3b](https://github.com/SatoshiPortal/boltz-rust/pull/91/commits/662ab3b85149560b4405b437a5cfe74ff49c39d6)

danielgranhao committed:
Bump crate version

roeierez reviewed (approved):
LGTM

i5hi commented:
Sorry about the delay in reviewing; closing up another release now and will take a look at this and test it tomorrow morning.

i5hi commented:
Getting this error when I run the tests:

```
---- btc_submarine_refund stdout ----

thread 'btc_submarine_refund' panicked at tests/test_framework/mod.rs:46:14:
called `Result::unwrap()` on an `Err` value: JsonRpc(Transport(SocketError(Os { code: 35, kind: WouldBlock, message: "Resource temporarily unavailable" })))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- btc_reverse_claim_relative_fee stdout ----
```

Similar error for :
test btc_submarine_refund ... FAILED
test btc_reverse_claim_relative_fee ... FAILED
test btc_reverse_claim_size ... FAILED
test btc_reverse_claim ... FAILED
test btc_submarine_refund_size ... FAILED
test btc_submarine_refund_relative_fee ... FAILED

danielgranhao commented:
> Getting this error when I run the tests:
> 
> ```
> ---- btc_submarine_refund stdout ----
> 
> thread 'btc_submarine_refund' panicked at tests/test_framework/mod.rs:46:14:
> called `Result::unwrap()` on an `Err` value: JsonRpc(Transport(SocketError(Os { code: 35, kind: WouldBlock, message: "Resource temporarily unavailable" })))
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 
> ---- btc_reverse_claim_relative_fee stdout ----
> ```
> 
> Similar error for : test btc_submarine_refund ... FAILED test btc_reverse_claim_relative_fee ... FAILED test btc_reverse_claim_size ... FAILED test btc_reverse_claim ... FAILED test btc_submarine_refund_size ... FAILED test btc_submarine_refund_relative_fee ... FAILED

@i5hi Could you please try to run the same tests on `trunk` (`cargo test`)? Those tests also fail on my machine (I have MacOS, not sure if you do too), but run fine in CI. Because for me, they already fail on `trunk`, I intended to leave out of this PR any potential fix. My assumption is that there is some issue with the `bitcoind` crate on MacOS.

I've opened an issue #93 for this.

i5hi commented:
> > Getting this error when I run the tests:
> > ```
> > ---- btc_submarine_refund stdout ----
> > 
> > thread 'btc_submarine_refund' panicked at tests/test_framework/mod.rs:46:14:
> > called `Result::unwrap()` on an `Err` value: JsonRpc(Transport(SocketError(Os { code: 35, kind: WouldBlock, message: "Resource temporarily unavailable" })))
> > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > 
> > ---- btc_reverse_claim_relative_fee stdout ----
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Similar error for : test btc_submarine_refund ... FAILED test btc_reverse_claim_relative_fee ... FAILED test btc_reverse_claim_size ... FAILED test btc_reverse_claim ... FAILED test btc_submarine_refund_size ... FAILED test btc_submarine_refund_relative_fee ... FAILED
> 
> @i5hi Could you please try to run the same tests on `trunk` (`cargo test`)? Those tests also fail on my machine (I have MacOS, not sure if you do too), but run fine in CI. Because for me, they already fail on `trunk`, I intended to leave out of this PR any potential fix. My assumption is that there is some issue with the `bitcoind` crate on MacOS.
> 
> I've opened an issue #93 for this.

Yes, I had tested it on a mac. Tested on Linux and its all good.

i5hi commented:
LGTM.

Thanks @danielgranhao !
